File: voteu-project/src/services/ConstituencyService.ts
import type { Constituency } from '@/interface/Constituency.ts';
import type { Party } from '@/interface/Party.ts';
import { apiFetch } from '@/services/api.ts'

/**
 * This service provides methods to interact with the backend API for retrieving data related to constituencies and their associated parties within the context of elections.
 */
export class ConstituencyServiceService {
  /**
   * Retrieves constituencies associated with a specific election by its ID.
   *
   * @param {string} electionId - The unique identifier for the election.
   * @return {Promise<Record<number, Constituency> | null>} A promise that resolves to a record of constituencies indexed by their IDs, or null if an error occurs.
   */
  static async getConstituenciesByElection(electionId: string): Promise<Record<number, Constituency> | null> {
    try {
      return await apiFetch<Record<number, Constituency>>(`/election/TK${electionId}/constituencies/compact`);
    } catch (error) {
      console.error(error);
      return null;
    }
  }

  /**
   * Fetches the votes for parties in a specific constituency for a given election.
   *
   * @param {string} electionId - The unique identifier of the election.
   * @param {string} constituencyId - The unique identifier of the constituency.
   * @return {Promise<Record<number, Party> | null>} A promise that resolves to a record mapping party IDs to party information, or null if the request fails.
   */
  static async getConstituencyPartyVotes(electionId: string, constituencyId: string): Promise<Record<number, Party> | null> {
    try {
      return await apiFetch<Record<number, Party>>(`/election/TK${electionId}/constituencies/${constituencyId}/parties`);
    } catch (error) {
      console.error(error);
      return null;
    }
  }
}

File: java/src/main/java/com/voteU/election/java/services/electiondata/memory/ConstituencyService.java
package com.voteU.election.java.services.electiondata.memory;
import com.voteU.election.java.dtoCompact.CompactConstituency;
import com.voteU.election.java.model.Constituency;
import com.voteU.election.java.model.Election;
import com.voteU.election.java.model.Party;
import org.springframework.stereotype.Service;
import java.util.*;

/**
 * Service class responsible for handling operations related to constituencies.
 * <p>
 * It fetches constituency data from the election service and provides methods
 * to access detailed and compact views of constituencies and their related parties.
 * </p>
 */
@Service
public class ConstituencyService {

    private final ElectionService electionService;

    /**
     * Constructor that injects the ElectionService dependency.
     *
     * @param electionService the election service to retrieve election data
     */
    public ConstituencyService(ElectionService electionService) {
        this.electionService = electionService;
    }

    /**
     * Retrieves all constituencies for a given election year.
     *
     * @param year the election year as a string
     * @return a map of constituency IDs to Constituency objects, or null if not found
     */
    public Map<Integer, Constituency> getConstituenciesByYear(String year) {
        Election election = electionService.getElection(year);
        if (election == null){
            return null;
        }
        Map<Integer, Constituency> constituencies = election.getConstituencies();
        if (constituencies == null || constituencies.isEmpty()) {
            System.out.println("No constituencies found for election year: " + year);
            return null;
        }
        return constituencies;
    }

    /**
     * Retrieves a compact representation of all constituencies for a given election year.
     *
     * @param year the election year as a string
     * @return a map of constituency IDs to CompactConstituency DTOs, or null if not found
     */
    public Map<Integer, CompactConstituency> getConstituenciesByYearCompact(String year) {
        Election election = electionService.getElection(year);
        if (election == null){
            return null;
        }
        Map<Integer, Constituency> constituencies = election.getConstituencies();
        if (constituencies == null || constituencies.isEmpty()) {
            System.out.println("No constituencies found for election year: " + year);
            return null;
        }
        Map<Integer, CompactConstituency> compactConstituencyMap = new HashMap<>();
        for (Constituency constituency : constituencies.values()) {
            CompactConstituency compactConstituency = new CompactConstituency(constituency.getId(), constituency.getName());
            compactConstituencyMap.put(compactConstituency.getId(), compactConstituency);
        }
        return compactConstituencyMap;
    }

    /**
     * Retrieves a specific constituency by its ID for a given election year.
     *
     * @param year the election year as a string
     * @param constituencyId the ID of the constituency
     * @return the Constituency object, or null if not found
     */
    public Constituency getConstituencyById(String year, int constituencyId) {
        Map<Integer, Constituency> constituencies = getConstituenciesByYear(year);
        if (constituencies == null){
            return null;
        }
        return constituencies.get(constituencyId);
    }

    /**
     * Retrieves all parties associated with a specific constituency for a given election year.
     *
     * @param year the election year as a string
     * @param constituencyId the ID of the constituency
     * @return a map of party IDs to Party objects, or null if not found
     */
    public Map<Integer, Party> getPartiesByConstituencyId(String year, int constituencyId) {
        Constituency constituency = getConstituencyById(year, constituencyId);
        if (constituency == null){
            return null;
        }
        return constituency.getParties();
    }
}
